File: frontend/src/lib/api/scripts.ts
/**
 * 스크립트 관련 API 서비스
 *
 * 백엔드 /api/v1/scripts 엔드포인트와 연동
 */

import { getApiClient, ApiError } from "./client.js";
import type { Script, Sentence } from "$lib/types/script.js";

// 백엔드 응답 타입 정의
interface BackendSentence {
  id: string;
  text: string;
  reading?: string; // 후리가나
  translation: string;
  start_time: number; // 시작 시간 (초)
  end_time: number; // 종료 시간 (초)
  difficulty_level: string;
}

interface BackendScript {
  id: string;
  title: string;
  description: string;
  audio_url: string;
  thumbnail_url?: string;
  duration: number; // 총 재생 시간 (초)
  difficulty_level: string;
  category: string; // news, anime, podcast, etc.
  language: string;
  created_at: string;
  sentences: BackendSentence[];
}

interface PlaybackProgress {
  script_id: string;
  current_time: number; // 현재 재생 시간 (초)
  completed_sentences: string[]; // 완료한 문장 ID 목록
  last_played: string; // ISO 문자열
}

// API 요청/응답 타입
interface GetScriptsParams {
  category?: string;
  difficulty?: string;
  limit?: number;
  offset?: number;
}

interface UpdateProgressRequest {
  current_time: number;
  completed_sentences: string[];
}

interface GetProgressResponse {
  script_id: string;
  current_time: number;
  completed_sentences: string[];
  last_played: string;
}

// 타입 변환 함수들
function convertBackendSentenceToFrontend(
  backendSentence: BackendSentence,
  orderIndex: number
): Sentence {
  return {
    id: backendSentence.id,
    content: backendSentence.text,
    orderIndex,
    metadata: {
      reading: backendSentence.reading,
      translation: backendSentence.translation,
      startTime: backendSentence.start_time,
      endTime: backendSentence.end_time,
      difficultyLevel: backendSentence.difficulty_level,
    },
  };
}

function convertBackendScriptToFrontend(backendScript: BackendScript): Script {
  const sentences = backendScript.sentences.map((sentence, index) =>
    convertBackendSentenceToFrontend(sentence, index)
  );

  return {
    id: backendScript.id,
    title: backendScript.title,
    description: backendScript.description,
    language: backendScript.language,
    sentences,
    mappings: [], // 매핑은 별도 API에서 로드
    metadata: {
      audioUrl: backendScript.audio_url,
      thumbnailUrl: backendScript.thumbnail_url,
      duration: backendScript.duration,
      difficultyLevel: backendScript.difficulty_level,
      category: backendScript.category,
    },
    createdAt: backendScript.created_at,
    updatedAt: backendScript.created_at, // 백엔드에서 updatedAt이 없으므로 createdAt 사용
  };
}

class ScriptsApiService {
  private client = getApiClient();

  /**
   * 스크립트 목록 조회
   */
  async getScripts(params: GetScriptsParams = {}): Promise<Script[]> {
    try {
      const queryParams = new URLSearchParams();

      if (params.category) queryParams.append("category", params.category);
      if (params.difficulty)
        queryParams.append("difficulty", params.difficulty);
      if (params.limit) queryParams.append("limit", params.limit.toString());
      if (params.offset) queryParams.append("offset", params.offset.toString());

      const queryString = queryParams.toString();
      const endpoint = `/scripts${queryString ? `?${queryString}` : ""}`;

      const backendScripts: BackendScript[] = await this.client.get(endpoint);

      return backendScripts.map(convertBackendScriptToFrontend);
    } catch (error) {
      console.error("Failed to fetch scripts:", error);
      throw error;
    }
  }

  /**
   * 특정 스크립트 상세 조회
   */
  async getScript(scriptId: string): Promise<Script> {
    try {
      const backendScript: BackendScript = await this.client.get(
        `/scripts/${scriptId}`
      );

      return convertBackendScriptToFrontend(backendScript);
    } catch (error) {
      console.error(`Failed to fetch script ${scriptId}:`, error);
      throw error;
    }
  }

  /**
   * 스크립트의 문장 목록만 조회
   */
  async getScriptSentences(scriptId: string): Promise<Sentence[]> {
    try {
      const backendSentences: BackendSentence[] = await this.client.get(
        `/scripts/${scriptId}/sentences`
      );

      return backendSentences.map((sentence, index) =>
        convertBackendSentenceToFrontend(sentence, index)
      );
    } catch (error) {
      console.error(`Failed to fetch sentences for script ${scriptId}:`, error);
      throw error;
    }
  }

  /**
   * 재생 진행률 업데이트
   */
  async updatePlaybackProgress(
    scriptId: string,
    progress: UpdateProgressRequest
  ): Promise<void> {
    try {
      const progressData: PlaybackProgress = {
        script_id: scriptId,
        current_time: progress.current_time,
        completed_sentences: progress.completed_sentences,
        last_played: new Date().toISOString(),
      };

      await this.client.post(`/scripts/${scriptId}/progress`, progressData);
    } catch (error) {
      console.error(`Failed to update progress for script ${scriptId}:`, error);
      // 진행률 업데이트 실패는 사용자 경험을 크게 해치지 않으므로 조용히 로그만 남김
      // throw하지 않음
    }
  }

  /**
   * 재생 진행률 조회
   */
  async getPlaybackProgress(
    scriptId: string
  ): Promise<GetProgressResponse | null> {
    try {
      const progress: GetProgressResponse = await this.client.get(
        `/scripts/${scriptId}/progress`
      );

      return progress;
    } catch (error) {
      if (error instanceof ApiError && error.status === 404) {
        // 진행률이 없는 경우 null 반환
        return null;
      }

      console.error(`Failed to fetch progress for script ${scriptId}:`, error);
      throw error;
    }
  }

  /**
   * 카테고리 목록 조회
   */
  async getCategories(): Promise<string[]> {
    try {
      const response: { categories: string[] } = await this.client.get(
        "/scripts/categories"
      );

      return response.categories || [];
    } catch (error) {
      console.error("Failed to fetch categories:", error);
      throw error;
    }
  }

  /**
   * 스크립트 북마크 추가
   */
  async bookmarkScript(scriptId: string): Promise<void> {
    try {
      await this.client.post(`/scripts/${scriptId}/bookmark`);
    } catch (error) {
      console.error(`Failed to bookmark script ${scriptId}:`, error);
      throw error;
    }
  }

  /**
   * 스크립트 북마크 제거
   */
  async removeBookmark(scriptId: string): Promise<void> {
    try {
      await this.client.delete(`/scripts/${scriptId}/bookmark`);
    } catch (error) {
      console.error(`Failed to remove bookmark for script ${scriptId}:`, error);
      throw error;
    }
  }

  /**
   * 스크립트 검색 (제목, 설명으로)
   */
  async searchScripts(
    query: string,
    params: GetScriptsParams = {}
  ): Promise<Script[]> {
    try {
      const searchParams = {
        ...params,
        q: query,
      };

      const queryParams = new URLSearchParams();
      Object.entries(searchParams).forEach(([key, value]) => {
        if (value !== undefined) {
          queryParams.append(key, value.toString());
        }
      });

      const endpoint = `/scripts/search?${queryParams.toString()}`;
      const backendScripts: BackendScript[] = await this.client.get(endpoint);

      return backendScripts.map(convertBackendScriptToFrontend);
    } catch (error) {
      console.error("Failed to search scripts:", error);
      throw error;
    }
  }
}

// 싱글톤 인스턴스
let scriptsApiInstance: ScriptsApiService | null = null;

/**
 * 스크립트 API 서비스 인스턴스 가져오기
 */
export function getScriptsApi(): ScriptsApiService {
  if (!scriptsApiInstance) {
    scriptsApiInstance = new ScriptsApiService();
  }
  return scriptsApiInstance;
}

/**
 * 스크립트 API 서비스 인스턴스 설정 (테스트용)
 */
export function setScriptsApi(api: ScriptsApiService): void {
  scriptsApiInstance = api;
}

// 기본 export
export const scriptsApi = getScriptsApi();

File: backend/app/api/v1/endpoints/scripts.py
"""
스크립트 관련 API 엔드포인트

라디오 스크립트 조회, 문장 싱크, 재생 관리를 담당합니다.
"""

from fastapi import APIRouter, Depends, HTTPException, status, Query
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
import logging

from app.core.auth import get_current_user
from app.models.user import User
from app.core.dependencies import get_database

logger = logging.getLogger(__name__)

router = APIRouter()

# 모델 정의
class Sentence(BaseModel):
    """문장 모델"""
    id: str
    text: str
    reading: Optional[str] = None     # 후리가나
    translation: str
    start_time: float                 # 시작 시간 (초)
    end_time: float                   # 종료 시간 (초)
    difficulty_level: str = "beginner"  # beginner, intermediate, advanced

class Script(BaseModel):
    """스크립트 모델"""
    id: str
    title: str
    description: str
    audio_url: str
    thumbnail_url: Optional[str] = None
    duration: int                     # 총 재생 시간 (초)
    difficulty_level: str = "beginner"
    category: str                     # news, anime, podcast, etc.
    language: str = "japanese"
    created_at: datetime
    sentences: List[Sentence]

class PlaybackProgress(BaseModel):
    """재생 진행률 모델"""
    script_id: str
    current_time: float               # 현재 재생 시간 (초)
    completed_sentences: List[str]    # 완료한 문장 ID 목록
    last_played: datetime

@router.get("/")
async def get_scripts(
    search: Optional[str] = Query(None, description="검색어"),
    category: Optional[str] = Query(None, description="카테고리 필터"),
    level: Optional[str] = Query(None, description="난이도 필터"),
    page: int = Query(1, ge=1, description="페이지 번호"),
    limit: int = Query(20, ge=1, le=100, description="페이지 크기")
):
    """
    스크립트 목록 조회
    
    카테고리, 난이도 등으로 필터링된 스크립트 목록을 반환합니다.
    """
    try:
        # Supabase에서 스크립트 목록 조회
        db = await get_database()
        
        # 기본 필터링 조건
        query = db.client.from_("scripts").select("*")
        
        # 검색어가 있으면 제목이나 설명에서 검색
        if search:
            query = query.or_(f"title.ilike.%{search}%,description.ilike.%{search}%")
        
        # 카테고리 필터
        if category:
            query = query.eq("category", category)
        
        # 레벨 필터  
        if level:
            query = query.eq("difficulty_level", level)
        
        # 정렬 및 페이징
        offset = (page - 1) * limit
        query = query.order("created_at", desc=True).range(offset, offset + limit - 1)
        
        result = query.execute()
        
        if not result.data:
            # 데이터가 없을 경우 더미 데이터 반환 (개발용)
            logger.info("DB에 스크립트가 없음. 더미 데이터 반환")
            # ... existing code ...
            sample_sentences = [
                Sentence(
                    id="sent_1",
                    text="今日は良い天気ですね。",
                    reading="きょうはいいてんきですね。",
                    translation="오늘은 좋은 날씨네요.",
                    start_time=0.0,
                    end_time=3.5,
                    difficulty_level="beginner"
                )
            ]
            
            mock_scripts = [
                Script(
                    id="script_1",
                    title="NHK 뉴스 - 오늘의 날씨",
                    description="일본 전국의 오늘 날씨를 전해드립니다.",
                    audio_url="https://example.com/audio/news_weather.mp3",
                    thumbnail_url="https://example.com/thumbnails/news.jpg",
                    duration=120,
                    difficulty_level="beginner",
                    category="news",
                    language="japanese",
                    created_at=datetime.utcnow(),
                    sentences=sample_sentences
                )
            ]
            
            return {
                "scripts": mock_scripts,
                "total": len(mock_scripts),
                "page": page,
                "limit": limit,
                "has_next": False
            }
        
        # 총 개수 조회 (페이징용)
        total_query = db.client.from_("scripts").select("id", count="exact")
        if search:
            total_query = total_query.or_(f"title.ilike.%{search}%,description.ilike.%{search}%")
        if category:
            total_query = total_query.eq("category", category)
        if level:
            total_query = total_query.eq("difficulty_level", level)
            
        total_result = total_query.execute()
        total_count = total_result.count or 0
        
        logger.info(f"스크립트 목록 조회 성공: {len(result.data)}개 (전체 {total_count}개)")
        
        return {
            "scripts": result.data,
            "total": total_count,
            "page": page,
            "limit": limit,
            "has_next": (page * limit) < total_count
        }
        
    except Exception as e:
        logger.error(f"스크립트 목록 조회 실패: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="스크립트 목록을 불러올 수 없습니다"
        )

@router.get("/{script_id}", response_model=Script)
async def get_script(script_id: str):
    """
    특정 스크립트 상세 조회
    
    스크립트 ID로 상세 정보와 모든 문장을 반환합니다.
    """
    try:
        # Supabase에서 스크립트 조회
        db = await get_database()
        
        # 스크립트 기본 정보 조회
        script_result = db.client.from_("scripts").select("*").eq("id", script_id).execute()
        
        if not script_result.data:
            # 개발용 더미 데이터 반환
            if script_id == "script_1":
                logger.info(f"DB에 스크립트 없음. 더미 데이터 반환: {script_id}")
                # ... existing dummy data code ...
                sample_sentences = [
                    Sentence(
                        id="sent_1",
                        text="今日は良い天気ですね。",
                        reading="きょうはいいてんきですね。",
                        translation="오늘은 좋은 날씨네요.",
                        start_time=0.0,
                        end_time=3.5,
                        difficulty_level="beginner"
                    )
                ]
                
                return Script(
                    id=script_id,
                    title="NHK 뉴스 - 오늘의 날씨",
                    description="일본 전국의 오늘 날씨를 전해드립니다.",
                    audio_url="https://example.com/audio/news_weather.mp3",
                    thumbnail_url="https://example.com/thumbnails/news.jpg",
                    duration=120,
                    difficulty_level="beginner",
                    category="news",
                    language="japanese",
                    created_at=datetime.utcnow(),
                    sentences=sample_sentences
                )
            else:
                logger.warning(f"존재하지 않는 스크립트 요청: {script_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="스크립트를 찾을 수 없습니다."
                )
        
        script_data = script_result.data[0]
        
        # 관련 문장들 조회
        sentences_result = db.client.from_("sentences").select("*").eq("script_id", script_id).order("start_time").execute()
        
        logger.info(f"스크립트 조회 성공: {script_id}")
        
        return Script(
            **script_data,
            sentences=sentences_result.data or []
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"스크립트 조회 실패: {script_id}, 오류: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="스크립트를 불러올 수 없습니다"
        )

@router.get("/{script_id}/sentences", response_model=List[Sentence])
async def get_script_sentences(script_id: str):
    """
    스크립트의 문장 목록 조회
    
    특정 스크립트의 모든 문장을 타임스탬프와 함께 반환합니다.
    """
    # TODO: 실제 DB에서 문장 목록 조회
    
    # 임시 문장 데이터
    return [
        Sentence(
            id="sent_1",
            text="今日は良い天気ですね。",
            reading="きょうはいいてんきですね。",
            translation="오늘은 좋은 날씨네요.",
            start_time=0.0,
            end_time=3.5,
            difficulty_level="beginner"
        ),
        Sentence(
            id="sent_2",
            text="明日雨が降るかもしれません。",
            reading="あしたあめがふるかもしれません。",
            translation="내일 비가 올지도 모릅니다.",
            start_time=4.0,
            end_time=8.2,
            difficulty_level="intermediate"
        )
    ]

@router.post("/{script_id}/progress")
async def update_playback_progress(
    script_id: str,
    progress: PlaybackProgress,
    current_user: User = Depends(get_current_user)
):
    """
    재생 진행률 업데이트
    
    사용자의 스크립트 재생 진행률을 저장합니다.
    """
    try:
        # TODO: 실제 DB에 사용자별 재생 진행률 저장
        
        logger.info(f"✅ 재생 진행률 업데이트 성공: {current_user.email}, 스크립트: {script_id}")
        
        return {
            "message": "재생 진행률이 저장되었습니다.",
            "script_id": script_id,
            "user_id": str(current_user.id),
            "current_time": progress.current_time,
            "completed_sentences": len(progress.completed_sentences),
            "updated_at": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"❌ 재생 진행률 업데이트 중 서버 에러: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="재생 진행률 업데이트 중 오류가 발생했습니다."
        )

@router.get("/{script_id}/progress")
async def get_playback_progress(
    script_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    재생 진행률 조회
    
    사용자의 특정 스크립트 재생 진행률을 반환합니다.
    """
    try:
        # TODO: 실제 DB에서 사용자별 진행률 조회
        
        logger.info(f"✅ 재생 진행률 조회 성공: {current_user.email}, 스크립트: {script_id}")
        
        # 임시 진행률 데이터
        return PlaybackProgress(
            script_id=script_id,
            current_time=45.5,
            completed_sentences=["sent_1"],
            last_played=datetime.utcnow()
        )
        
    except Exception as e:
        logger.error(f"❌ 재생 진행률 조회 중 서버 에러: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="재생 진행률 조회 중 오류가 발생했습니다."
        )

@router.get("/categories/")
async def get_categories():
    """
    스크립트 카테고리 목록 조회
    
    사용 가능한 모든 카테고리를 반환합니다.
    """
    return {
        "categories": [
            {"id": "news", "name": "뉴스", "count": 25},
            {"id": "anime", "name": "애니메이션", "count": 18},
            {"id": "podcast", "name": "팟캐스트", "count": 12},
            {"id": "drama", "name": "드라마", "count": 8},
            {"id": "music", "name": "음악", "count": 15}
        ]
    }

@router.post("/{script_id}/bookmark")
async def bookmark_script(
    script_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    스크립트 북마크 추가
    
    사용자의 북마크 목록에 스크립트를 추가합니다.
    """
    try:
        # TODO: 실제 북마크 추가 로직 구현
        
        logger.info(f"✅ 스크립트 북마크 추가 성공: {current_user.email}, 스크립트: {script_id}")
        
        return {
            "message": f"스크립트 {script_id}가 북마크에 추가되었습니다.",
            "user_id": str(current_user.id),
            "script_id": script_id,
            "bookmarked_at": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"❌ 스크립트 북마크 추가 중 서버 에러: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="북마크 추가 중 오류가 발생했습니다."
        )

@router.delete("/{script_id}/bookmark")
async def remove_bookmark(
    script_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    스크립트 북마크 제거
    
    사용자의 북마크 목록에서 스크립트를 제거합니다.
    """
    try:
        # TODO: 실제 북마크 제거 로직 구현
        
        logger.info(f"✅ 스크립트 북마크 제거 성공: {current_user.email}, 스크립트: {script_id}")
        
        return {
            "message": f"스크립트 {script_id}가 북마크에서 제거되었습니다.",
            "user_id": str(current_user.id),
            "script_id": script_id,
            "removed_at": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"❌ 스크립트 북마크 제거 중 서버 에러: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="북마크 제거 중 오류가 발생했습니다."
        ) 